update-vo-config/update-vo-config

```diff
@@ -0,0 +1,507 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import argparse
+from operator import itemgetter
+import os
+import re
+import sys
+import xml.etree.ElementTree as ET
+from OpenSSL import crypto
+import urllib2
+
+SW_MANAGER_ROLES = ['/Role=lcgadmin', '/admin', '/Role=swadmin',
+                    '/Role=sgmadmin', '/Role=sgm', '/Role=SoftwareManager',
+                    '/Role=VO-Software-Manager', '/Role=SW-Admin']
+
+PRODUCTION_ROLES = ['/Role=production', '/Role=Production', '/Role=prod',
+                    '/Role=ProductionManager']
+
+PILOTE_ROLES = ['/Role=pilot']
+
+FQAN_TYPE_MAPPING = {
+    'pilot': 'pilot',
+    'production': 'p',
+    'SW manager': 's'
+}
+
+VO_CARD_URI = "http://operations-portal.egi.eu/xml/voIDCard/public/all/true"
+CERT_DIR = 'vo/certs'
+PARAMS_DIR = 'vo/params'
+SITE_PARAMS_DIR = 'vo/site'
+# The base XML element of the VO Card
+ROOT_ELEM = "VoDump"
+ID_CARD_ELEM = "IDCard"
+ID_ATTR = "CIC_ID"
+
+DESCRIPTION = "Configure VO-related PAN templates"
+VERSION = "0.1"
+
+
+class VOMSEndpoint(object):
+    """A class representing a VOMS server endpoint (used by a specific VO)
+    """
+    def __init__(self):
+        self.hostname = None
+        self.https_port = 8443
+        self.vomses_port = 0
+        self.old_cert_retrieved = False
+        self.cert = None
+        self.old_cert = None
+        self.is_admin = True
+
+    def get_cert(self):
+        """Check the certificate
+        """
+        pass
+
+    def set_is_admin(self, is_voms_admin_server):
+        """Set the value of the is_admin attribute
+        """
+        self.is_admin = (is_voms_admin_server == '1')
+
+
+def get_clean_fqan(fqan, vo_name):
+    """Cleanup the fqan
+    """
+    if not fqan:
+        return ""
+    # Remove leading/trailiong spaces if any added by mistake...
+    clean_fqan = fqan.strip()
+    # Remove /Role=NULL if specified in VO ID card
+    clean_fqan = re.sub(re.compile("/Role=NULL$"), "", clean_fqan)
+    # If the relative FQAN corresponds to the VO name without any role,
+    # set it to an empty string. An empty FQAN must not be added to the
+    # FQAN list as it is not a real FQAN.
+    if len(re.sub(re.compile("^/" + vo_name), "", clean_fqan)) == 0:
+        clean_fqan = ""
+    return clean_fqan
+
+
+def get_fqan_type(fqan, vo_name):
+    """ Check if fqan has a predefined role
+    """
+    clean_fqan = re.sub(re.compile("^/" + vo_name), "", fqan)
+    if clean_fqan in SW_MANAGER_ROLES:
+        return "SW manager"
+    elif clean_fqan in PRODUCTION_ROLES:
+        return "production"
+    elif clean_fqan in PILOTE_ROLES:
+        return "pilot"
+    else:
+        return None
+
+
+def java_string_hashcode(string):
+    """Python implementation of the Java String.hashCode function
```

Comment:
It is not my original work. It is taken from:
http://garage.pimentech.net/libcommonPython_src_python_libcommon_javastringhashcode/

Reply:
Should I add the link to the comment?

Reply:
Yes, I think that would be the right thing to do.
I believe the code is "Copyright 2010 PimenTech SARL", so that should most likely be added to the comment as well.

Reply:
The copyright has been added.